packages/drivers/odsp-driver/src/epochTracker.ts

```diff
@@ -266,17 +266,37 @@ export class EpochTrackerWithRedemption extends EpochTracker {
             // Similar, if getVersions failed, we should not do any further storage calls.
             // So treesLatest is the only call that can have parallel joinSession request.
             if (fetchType === "treesLatest") {
-                this.treesLatestDeferral.reject(error);
+                this.treesLatestDeferral.deferral.reject(error);
             }
-            if (fetchType !== "joinSession" || error.statusCode !== 404 || !completed) {
+            if (fetchType !== "joinSession" || error.statusCode !== 404 || completed) {
                 throw error;
             }
         }
 
         // It is joinSession failing with 404.
         // Repeat after waiting for treeLatest succeeding (of fail if it fails).
         // No special handling after first call - if file has been deleted, then it's game over.
-        await this.treesLatestDeferral.promise;
+        await this.treesLatestDeferral.addListenerWithCallback();
```

Comment:
I do not think that name is reflecting what it does - it does not add listener.
I'd try to make it as transparent as possible.
I.e. I'd make DeferralWithCallback implement those same APIs that Deferred does, such that it can be swapped in place without need to know or care about new names.

Reply:
I have updated the code with a more appropriate name. The reason I hadn't extended the Deferred class when setting up the new DeferralWithCallback is the API just provides the promise as a variable to be listened on. We need a function call here to initiate the callback when we do this await